File: ts/tests/unit/validators.test.ts
import { describe, it, expect } from 'vitest';
import { validateAIDFormat, validateGroupAIDFormat, validateGroupIDFormat, ValidationError } from '../../src/index.js';

describe('validators', () => {
  describe('validateAIDFormat', () => {
    it('should accept valid AID', () => {
      expect(validateAIDFormat('alice.aid.com')).toBe('alice.aid.com');
      expect(validateAIDFormat('bob_123.example.org')).toBe('bob_123.example.org');
      expect(validateAIDFormat('test-user.domain.co')).toBe('test-user.domain.co');
    });

    it('should normalize to lowercase', () => {
      expect(validateAIDFormat('Alice.AID.COM')).toBe('alice.aid.com');
      expect(validateAIDFormat('BOB1.Example.Org')).toBe('bob1.example.org');
    });

    it('should reject empty AID', () => {
      expect(() => validateAIDFormat('')).toThrow(ValidationError);
      expect(() => validateAIDFormat(null)).toThrow(ValidationError);
      expect(() => validateAIDFormat(undefined)).toThrow(ValidationError);
    });

    it('should reject AID without issuer', () => {
      expect(() => validateAIDFormat('alice')).toThrow(ValidationError);
      expect(() => validateAIDFormat('alice.')).toThrow(ValidationError);
    });

    it('should reject AID starting with guest', () => {
      expect(() => validateAIDFormat('guest123.aid.com')).toThrow(ValidationError);
      expect(() => validateAIDFormat('guestuser.example.org')).toThrow(ValidationError);
    });

    it('should reject AID name starting with hyphen', () => {
      expect(() => validateAIDFormat('-alice.aid.com')).toThrow(ValidationError);
    });

    it('should reject AID name too short', () => {
      expect(() => validateAIDFormat('ab.aid.com')).toThrow(ValidationError);
      expect(() => validateAIDFormat('bob.aid.com')).toThrow(ValidationError);
    });

    it('should accept AID name at the 4-character minimum', () => {
      expect(validateAIDFormat('bob1.aid.com')).toBe('bob1.aid.com');
    });

    it('should reject invalid characters in name', () => {
      expect(() => validateAIDFormat('alice@bob1.aid.com')).toThrow(ValidationError);
      expect(() => validateAIDFormat('alice bob1.aid.com')).toThrow(ValidationError);
    });

    it('should reject invalid domain', () => {
      expect(() => validateAIDFormat('alice.-invalid')).toThrow(ValidationError);
      expect(() => validateAIDFormat('alice.invalid-')).toThrow(ValidationError);
    });
  });

  describe('validateGroupIDFormat', () => {
    it('should accept legacy format', () => {
      expect(validateGroupIDFormat('g-abcd1234')).toBe('g-abcd1234');
      expect(validateGroupIDFormat('g-test')).toBe('g-test');
    });

    it('should accept new base format', () => {
      expect(validateGroupIDFormat('abcde')).toBe('abcde');
      expect(validateGroupIDFormat('12345678')).toBe('12345678');
    });

    it('should accept group name format', () => {
      expect(validateGroupIDFormat('team_alpha')).toBe('team_alpha');
      expect(validateGroupIDFormat('group-beta')).toBe('group-beta');
    });

    it('should accept canonical format', () => {
      expect(validateGroupIDFormat('group.aid.com/g-test')).toBe('g-test.aid.com');
      expect(validateGroupIDFormat('group.example.org/mygroup')).toBe('mygroup.example.org');
    });

    it('should accept @ format', () => {
      expect(validateGroupIDFormat('g-test@aid.com')).toBe('g-test.aid.com');
      expect(validateGroupIDFormat('mygroup@example.org')).toBe('mygroup.example.org');
    });

    it('should accept dot format', () => {
      expect(validateGroupIDFormat('g-test.aid.com')).toBe('g-test.aid.com');
      expect(validateGroupIDFormat('mygroup.example.org')).toBe('mygroup.example.org');
    });

    it('should normalize to lowercase', () => {
      expect(validateGroupIDFormat('G-TEST')).toBe('g-test');
      expect(validateGroupIDFormat('MyGroup.EXAMPLE.ORG')).toBe('mygroup.example.org');
    });

    it('should return target group_aid format for legacy inputs', () => {
      expect(validateGroupAIDFormat('room-123.agentid.pub')).toBe('room-123.agentid.pub');
      expect(validateGroupAIDFormat('group.agentid.pub/room-123')).toBe('room-123.agentid.pub');
      expect(validateGroupAIDFormat('room-123@agentid.pub')).toBe('room-123.agentid.pub');
      expect(validateGroupAIDFormat('g-abc123', { localIssuer: 'agentid.pub' })).toBe('g-abc123.agentid.pub');
      expect(validateGroupAIDFormat('group.pub/room-123@agentid')).toBe('room-123.agentid.pub');
    });

    it('should reject empty group_id', () => {
      expect(() => validateGroupIDFormat('')).toThrow(ValidationError);
      expect(() => validateGroupIDFormat(null)).toThrow(ValidationError);
      expect(() => validateGroupIDFormat(undefined)).toThrow(ValidationError);
    });

    it('should reject invalid base format', () => {
      expect(() => validateGroupIDFormat('ab')).toThrow(ValidationError); // 太短（只有2个字符）
      expect(() => validateGroupIDFormat('123')).toThrow(ValidationError); // 太短（只有3个字符）
    });

    it('should reject invalid domain', () => {
      expect(() => validateGroupIDFormat('g-test.-invalid')).toThrow(ValidationError);
      expect(() => validateGroupIDFormat('mygroup@invalid-')).toThrow(ValidationError);
    });
  });
});

File: go/validators_test.go
package aun

import (
	"strings"
	"testing"
)

func TestValidateAIDFormat(t *testing.T) {
	tests := []struct {
		name      string
		aid       any
		paramName string
		wantErr   bool
		errMsg    string
	}{
		// 正常情况
		{name: "valid aid", aid: "alice.aid.com", paramName: "aid", wantErr: false},
		{name: "valid aid with underscore", aid: "alice_bob.aid.com", paramName: "aid", wantErr: false},
		{name: "valid aid with dash", aid: "alice-bob.aid.com", paramName: "aid", wantErr: false},
		{name: "valid aid 4 chars", aid: "alex.aid.com", paramName: "aid", wantErr: false},
		{name: "valid aid uppercase normalized", aid: "Alice.Aid.Com", paramName: "aid", wantErr: false},

		// 空值
		{name: "nil aid", aid: nil, paramName: "aid", wantErr: true, errMsg: "cannot be empty"},
		{name: "empty string", aid: "", paramName: "aid", wantErr: true, errMsg: "cannot be empty"},
		{name: "whitespace only", aid: "   ", paramName: "aid", wantErr: true, errMsg: "cannot be empty"},

		// 缺少域名
		{name: "no issuer", aid: "alice", paramName: "aid", wantErr: true, errMsg: "must be in format"},
		{name: "trailing dot", aid: "alice.", paramName: "aid", wantErr: true, errMsg: "issuer part cannot be empty"},

		// name 部分不合法
		{name: "name too short", aid: "ali.aid.com", paramName: "aid", wantErr: true, errMsg: "must be 4-64 characters"},
		{name: "name starts with dash", aid: "-alice.aid.com", paramName: "aid", wantErr: true, errMsg: "cannot start with '-'"},
		{name: "name starts with guest", aid: "guest123.aid.com", paramName: "aid", wantErr: true, errMsg: "cannot start with 'guest'"},
		{name: "name with invalid char", aid: "alice@bob.aid.com", paramName: "aid", wantErr: true, errMsg: "must be 4-64 characters"},
		{name: "name with space", aid: "alice bob.aid.com", paramName: "aid", wantErr: true, errMsg: "must be 4-64 characters"},

		// issuer 部分不合法
		{name: "invalid issuer", aid: "alice.-aid.com", paramName: "aid", wantErr: true, errMsg: "is not a valid domain"},
		{name: "issuer with space", aid: "alice.aid com", paramName: "aid", wantErr: true, errMsg: "is not a valid domain"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result, err := ValidateAIDFormat(tt.aid, tt.paramName)
			if tt.wantErr {
				if err == nil {
					t.Errorf("ValidateAIDFormat() expected error but got nil")
					return
				}
				if tt.errMsg != "" && !strings.Contains(err.Error(), tt.errMsg) {
					t.Errorf("ValidateAIDFormat() error = %v, want error containing %q", err, tt.errMsg)
				}
			} else {
				if err != nil {
					t.Errorf("ValidateAIDFormat() unexpected error = %v", err)
					return
				}
				if result == "" {
					t.Errorf("ValidateAIDFormat() returned empty string for valid input")
				}
				// 验证结果是小写的
				if result != strings.ToLower(result) {
					t.Errorf("ValidateAIDFormat() result not lowercase: %s", result)
				}
			}
		})
	}
}

func TestValidateGroupIDFormat(t *testing.T) {
	tests := []struct {
		name      string
		groupID   any
		paramName string
		wantErr   bool
		errMsg    string
	}{
		// 正常情况 - Legacy 格式
		{name: "legacy format", groupID: "g-abcd1234", paramName: "group_id", wantErr: false},
		{name: "legacy format min", groupID: "g-abcd", paramName: "group_id", wantErr: false},
		{name: "legacy format max", groupID: "g-abcd1234567890abcd1234567890ab", paramName: "group_id", wantErr: false},

		// 正常情况 - 新格式
		{name: "new format 5 chars", groupID: "abcde", paramName: "group_id", wantErr: false},
		{name: "new format longer", groupID: "abcde12345", paramName: "group_id", wantErr: false},

		// 正常情况 - Group name 格式
		{name: "group name", groupID: "mygroup", paramName: "group_id", wantErr: false},
		{name: "group name with underscore", groupID: "my_group", paramName: "group_id", wantErr: false},
		{name: "group name with dash", groupID: "my-group", paramName: "group_id", wantErr: false},

		// 正常情况 - 带域名
		{name: "with domain dot", groupID: "mygroup.aid.com", paramName: "group_id", wantErr: false},
		{name: "with domain @", groupID: "mygroup@aid.com", paramName: "group_id", wantErr: false},
		{name: "canonical format", groupID: "group.aid.com/mygroup", paramName: "group_id", wantErr: false},
		{name: "polluted canonical format", groupID: "group.pub/room-123@agentid", paramName: "group_id", wantErr: false},
		{name: "legacy with domain", groupID: "g-abcd1234.aid.com", paramName: "group_id", wantErr: false},

		// 空值
		{name: "nil group_id", groupID: nil, paramName: "group_id", wantErr: true, errMsg: "cannot be empty"},
		{name: "empty string", groupID: "", paramName: "group_id", wantErr: true, errMsg: "cannot be empty"},
		{name: "whitespace only", groupID: "   ", paramName: "group_id", wantErr: true, errMsg: "cannot be empty"},

		// base 部分不合法
		{name: "legacy too short", groupID: "g-a", paramName: "group_id", wantErr: true, errMsg: "must be one of"},
		{name: "too short 3 chars", groupID: "abc", paramName: "group_id", wantErr: true, errMsg: "must be one of"},
		{name: "name starts with dash", groupID: "-mygroup", paramName: "group_id", wantErr: true, errMsg: "must be one of"},
		{name: "name with invalid char", groupID: "my@group", paramName: "group_id", wantErr: true, errMsg: "must be one of"},
		{name: "name with space", groupID: "my group", paramName: "group_id", wantErr: true, errMsg: "must be one of"},

		// domain 部分不合法
		{name: "invalid domain", groupID: "mygroup.-aid.com", paramName: "group_id", wantErr: true, errMsg: "is not a valid domain"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result, err := ValidateGroupIDFormat(tt.groupID, tt.paramName)
			if tt.wantErr {
				if err == nil {
					t.Errorf("ValidateGroupIDFormat() expected error but got nil")
					return
				}
				if tt.errMsg != "" && !strings.Contains(err.Error(), tt.errMsg) {
					t.Errorf("ValidateGroupIDFormat() error = %v, want error containing %q", err, tt.errMsg)
				}
			} else {
				if err != nil {
					t.Errorf("ValidateGroupIDFormat() unexpected error = %v", err)
					return
				}
				if result == "" {
					t.Errorf("ValidateGroupIDFormat() returned empty string for valid input")
				}
				// 验证结果是小写的
				if result != strings.ToLower(result) {
					t.Errorf("ValidateGroupIDFormat() result not lowercase: %s", result)
				}
				if strings.HasPrefix(result, "group.") && strings.Contains(result, "/") {
					t.Errorf("ValidateGroupIDFormat() must return group_aid target format, got %s", result)
				}
			}
		})
	}
}

func TestValidateGroupAIDFormat(t *testing.T) {
	tests := []struct {
		raw         string
		localIssuer string
		want        string
	}{
		{raw: "room-123.agentid.pub", want: "room-123.agentid.pub"},
		{raw: "group.agentid.pub/room-123", want: "room-123.agentid.pub"},
		{raw: "room-123@agentid.pub", want: "room-123.agentid.pub"},
		{raw: "g-abc123", localIssuer: "agentid.pub", want: "g-abc123.agentid.pub"},
		{raw: "group.pub/room-123@agentid", want: "room-123.agentid.pub"},
	}
	for _, tt := range tests {
		t.Run(tt.raw, func(t *testing.T) {
			got, err := ValidateGroupAIDFormat(tt.raw, "group_aid", tt.localIssuer)
			if err != nil {
				t.Fatalf("ValidateGroupAIDFormat() error = %v", err)
			}
			if got != tt.want {
				t.Fatalf("ValidateGroupAIDFormat() = %q, want %q", got, tt.want)
			}
		})
	}
}
